File: app/src/main/java/com/example/marco/weather/Data/Location.java
package com.example.marco.weather.Data;

import com.google.gson.*;
import java.lang.reflect.Type;
import io.realm.RealmList;
import io.realm.RealmObject;

public class Location extends RealmObject {

    private int id;
    private String name;
    private String country;
    private long updateTime;
    private RealmList<Weather> forecast = new RealmList<>();

    public int getId() {
        return id;
    }

    private void setId(int id) {
        this.id = id;
    }

    public void setId(String id) {
        this.id = Integer.parseInt(id);
    }

    public String getName() {
        return name;
    }

    private void setName(String name) {
        this.name = name;
    }

    public String getCountry() {return country;}

    private void setCountry(String country) {
        this.country = country;
    }

    public long getUpdateTime() {
        return updateTime;
    }

    public void setUpdateTime(long updateTime) {
        this.updateTime = updateTime;
    }

    public Weather getForecast (int date) {
        return forecast.get(date);
    }

    public void setForecast (int date, Weather weather) {
        forecast.add(date, weather);
    }

    @Override
    public boolean equals(Object other) {
        return other != null && (other == this || other instanceof Location && id == ((Location) other).id);
    }
    @Override
    public String toString() {
        return name+"\n"+country+"\n"+id+"\n";
    }

    public static class CityDeserializer implements JsonDeserializer<Location> {

        @Override
        public Location deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
            Location location = new Gson().fromJson(json, Location.class);
            JsonObject jsonObject = json.getAsJsonObject();
            location.setId(jsonObject.get("Key").getAsInt());
            location.setName(jsonObject.get("LocalizedName").getAsString());
            location.setCountry(jsonObject.getAsJsonObject("Country").get("LocalizedName").getAsString());
            return location;
        }
    }
}

File: app/src/main/java/com/example/marco/weather/Data/Location.kt
package com.example.marco.weather.Data

import com.google.gson.*
import java.lang.reflect.Type
import io.realm.RealmList
import io.realm.RealmObject

open class Location(var id: Int = 0, var name: String = "", var country: String = "", var updateTime: Long = 0, var forecast: RealmList<Weather> = RealmList<Weather>()) : RealmObject() {

    fun getForecastAt(date: Int): Weather {
        return forecast[date]
    }

    fun setForecastAt(date: Int, weather: Weather) {
        forecast.add(date, weather)
    }

    class CityDeserializer : JsonDeserializer<Location> {

        override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Location {
            val location = Gson().fromJson<Location>(json, Location::class.java)
            val jsonObject = json.asJsonObject
            location.id = jsonObject.get("Key").asInt
            location.name = jsonObject.get("LocalizedName").asString
            location.country = jsonObject.getAsJsonObject("Country").get("LocalizedName").asString
            return location
        }
    }
}
